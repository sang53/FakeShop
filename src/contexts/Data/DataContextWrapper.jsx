import { useEffect, useState } from "react";
import Helpers from "../../Helper/Helpers";
import { Products } from "./DataContext";
import PropTypes from "prop-types";

export default function DataContextWrapper({ children }) {
  const [data, setData] = useState([]);

  useEffect(() => {
    async function getData() {
      const newData = await Helpers.getProduct();
      if (useData) setData(newData);
    }
    let useData = true;
    getData();

    return () => (useData = false);
  }, []);

  return <Products.Provider value={data}>{children}</Products.Provider>;
}

DataContextWrapper.propTypes = {
  children: PropTypes.element.isRequired,
};
