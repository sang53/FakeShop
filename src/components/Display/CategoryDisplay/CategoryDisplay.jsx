import { useParams } from "react-router";
import DisplayContainer from "../DisplayContainer/DisplayContainer";
import { Products } from "../../../contexts/Data/DataContext";
import { use } from "react";
import DisplayObj from "../DisplayObj/DisplayObj";

export default function CategoryDisplay() {
  const { category } = useParams();
  const displayItems = category
    ? DisplayObj.getCategoryProducts(use(Products), category)
    : DisplayObj.getAllCategories(use(Products));

  return (
    <>
      <h1>{category ? DisplayObj.capitalise(category) : "Categories"}</h1>
      <DisplayContainer displayItems={displayItems} />
    </>
  );
}
