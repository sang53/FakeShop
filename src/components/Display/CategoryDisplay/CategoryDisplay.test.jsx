import { render, screen } from "@testing-library/react";
import { describe, expect, test } from "vitest";
import { Products } from "../../../contexts/Data/DataContext";
import { MemoryRouter, Route, Routes } from "react-router";
import CategoryDisplay from "./CategoryDisplay";
import { mockProducts } from "../../../test/utils/mockProducts";

function renderCategories(value, category = "") {
  return render(
    <Products.Provider value={value}>
      <MemoryRouter initialEntries={[`/categories${category}`]}>
        <Routes>
          <Route path="/categories" element={<CategoryDisplay />} />
          <Route path="/categories/:category" element={<CategoryDisplay />} />
        </Routes>
      </MemoryRouter>
    </Products.Provider>
  );
}

describe("displays all categories correctly", () => {
  test("displays all categories", () => {
    renderCategories(mockProducts);
    expect(screen.getAllByRole("link")).toHaveLength(4);
  });

  test("all categories elements are displayed as expected", () => {
    const { container } = renderCategories(mockProducts.slice(0, 1));
    expect(screen.getByRole("heading")).toHaveTextContent(/categories/i);
    expect(container).toMatchSnapshot();
  });
});

describe("displays single category correctly", () => {
  test("displays single category", () => {
    renderCategories(mockProducts, "/men's clothing");
    expect(screen.getAllByRole("link")).toHaveLength(4);
  });
  test("single category elements are displayed as expected", () => {
    const { container } = renderCategories(
      mockProducts.slice(0, 1),
      "/men's clothing"
    );
    expect(screen.getByRole("heading")).toHaveTextContent(/men's clothing/i);
    expect(container).toMatchSnapshot();
  });
});
