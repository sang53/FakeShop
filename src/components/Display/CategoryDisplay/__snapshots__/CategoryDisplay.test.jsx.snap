// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`displays all categories correctly > all categories elements are displayed as expected 1`] = `
<div>
  <h1>
    Categories
  </h1>
  <div
    class="flex gap-1rem pad-1rem _displayContainer_acfe5e"
  >
    <a
      class="product"
      data-discover="true"
      href="/categories/men's clothing"
    >
      <div
        class="product-img"
      >
        <img
          alt="Men's Clothing"
          src="#"
        />
      </div>
      <div>
        <div
          id="title"
        >
          Men's Clothing
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`displays single category correctly > single category elements are displayed as expected 1`] = `
<div>
  <h1>
    Men's Clothing
  </h1>
  <div
    class="flex gap-1rem pad-1rem _displayContainer_acfe5e"
  >
    <a
      class="product"
      data-discover="true"
      href="/products/1"
    >
      <div
        class="product-img"
      >
        <img
          alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
          src="#"
        />
      </div>
      <div>
        <div
          id="title"
        >
          Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
        </div>
        <div>
          $109.95
        </div>
      </div>
    </a>
  </div>
</div>
`;
